### PR TITLE
Fix browser test flakiness from page JS load race conditions.

### LINKED ISSUE #1309: Browser test flake - failed program/question creation
**Describe the bug**
The latest attempt to deflake browser test #1305  didn't fully fix our prod promotion. We have issue when we create a new question or program, somehow the browser test still stays on the create form page.

**Error message**
Program creation:
```
FAIL browser: chromium src/admin_program_translation.test.ts (48.774 s)
  ● Admin can manage translations › create a question and add translations

    expect(received).toEqual(expected) // deep equality

    Expected: "All programs"
    Received: "New program information"

      15 |
      16 |   async expectAdminProgramsPage() {
    > 17 |     expect(await this.page.innerText('h1')).toEqual('All programs');
         |                                             ^
      18 |   }
      19 |
      20 |   async expectProgramExist(programName: string, description: string) {

      at AdminPrograms.<anonymous> (src/support/admin_programs.ts:17:45)
      at step (src/support/admin_programs.ts:33:23)
      at Object.next (src/support/admin_programs.ts:14:53)
      at fulfilled (src/support/admin_programs.ts:5:58)
```
Question creation:
```
FAIL browser: chromium src/admin_program_creation.test.ts
  ● Create program with enumerator and repeated questions › create program with enumerator and repeated questions

    expect(received).toEqual(expected) // deep equality

    Expected: "All Questions"
    Received: "New address question"

      16 |
      17 |   async expectAdminQuestionsPage() {
    > 18 |     expect(await this.page.innerText('h1')).toEqual('All Questions');
         |                                             ^
      19 |   }
      20 |
      21 |   async fillInQuestionBasics(questionName: string,

      at AdminQuestions.<anonymous> (src/support/admin_questions.ts:18:45)
      at step (src/support/admin_questions.ts:33:23)
      at Object.next (src/support/admin_questions.ts:14:53)
      at fulfilled (src/support/admin_questions.ts:5:58)
```

See full error message here https://github.com/seattle-uat/civiform/runs/2765182536?check_suite_focus=true

**Expected behavior**
The tests pass without errors, i.e. the creation succeeds and returns back to program/question list page.

**Additional context**
We recently found out that sometimes page may not be fully loaded (especially javascript) before the browser tests start interacting with the page. This *may* be also caused by such timing issue and adding additional wait guard *may* solve it. More investigation needed though.

### PR DESCRIPTION
**Pull request by leonwong**
### Description
Browser tests were flaky because they would start running while civiform JS would still be initializing (e.g., attaching event handlers). This PR adds some attributes for main.ts and modal.ts (the main causes of flakiness) to signal when they're done, and some browser test functions to wait for our pages to be ready and modals to be shown. Also tweaked a few selectors to make sure that playwright is waiting for DOM elements that signal when the page is actually ready for its next actions.

### Checklist
- [X] Created tests which fail without the change (if possible)
- [X] Extended the README / documentation, if necessary

### Issue(s)
Fixes #1309

### CODE REVIEW

**Comment by leonwong** [comment]
Note - will update the wiki with best practices after merging this PR.

**Review by ettengererin — APPROVED** [review]
This looks great Leon. Thanks for spending so much time on this fix, must feel good to see success : D 

Could you add a link/screenshot of tests passing in the PR description? It'd be helpful to document proof of flakiness resolution, so we can look back at it in the future if necessary.

**Comment by leonwong** [comment]
Additional notes:
- PR browser tests passed first time
- Screenshot of tests passing on my machine:

Screen Shot 2021-09-14 at 1 17 04 PM

While it's possible we may find other places where flakiness can occur (hard to catch them all esp as the code evolves), hopefully once we have some experience with this PR it will show that we have a clear understanding of the root causes, how to fix, and addresses the vast majority of current flakiness.

**Comment by bion** [comment]
🙌🏻

File: browser-test/src/end_to_end_enumerators.test.ts [hunk 1]
```diff
@@ -1,4 +1,5 @@
-import { startSession, loginAsAdmin, loginAsGuest, AdminQuestions, AdminPrograms, ApplicantQuestions, selectApplicantLanguage, endSession } from './support'
+import { waitForDebugger } from 'inspector';
```
Comment (bion) [inline browser-test/src/end_to_end_enumerators.test.ts:1]: nit: unused import

File: browser-test/src/support/admin_questions.ts [hunk 2]
```diff
@@ -11,14 +12,13 @@ export class AdminQuestions {
...
+    await waitForPageJsLoad(this.page);
     await this.expectAdminQuestionsPage();
   }
 
   async clickSubmitButtonAndNavigate(buttonText: string) {
-    await Promise.all([
-      this.page.waitForNavigation(),
-      this.page.click('button:has-text("' + buttonText + '")'),
-    ]);
+    await this.page.click('button:has-text("' + buttonText + '")');
```
Comment (bion) [inline browser-test/src/support/admin_questions.ts:20]: nit:
```suggestion
    await this.page.click(`button:has-text("${buttonText}")');
```

**Comment by bion** [comment]
Huge improvement for developer productivity, great work. Approved with a couple nits.

**Review by bion — APPROVED** [review]

File: browser-test/src/end_to_end_enumerators.test.ts (hunk 1 above)
Comment (leonwong) [reply browser-test/src/end_to_end_enumerators.test.ts]: removed

File: browser-test/src/support/admin_questions.ts (hunk 2 above)
Comment (leonwong) [reply browser-test/src/support/admin_questions.ts]: changed

**Review by dkatzz — APPROVED** [review]
This is great, thank you Leon!!

**Pull request merged**